### PR TITLE
Upload a document with metadata in single call

### PR DESCRIPTION
**Pull request by akorneyc**
This PR include Upload a document with metadata in single call

### CODE REVIEW

File: doc/dokuti_api.yml [hunk 1]
```diff
@@ -1903,6 +1911,21 @@ components:
         value:
           type: string
           description: The attribute value.
+    DocumentAttributeValueRequest:
```
Comment (melissapalmer) [inline doc/dokuti_api.yml]: How is this model different from `DocumentAttributeRequest` should we not be using the same model?

File: doc/dokuti_api.yml [hunk 2]
```diff
@@ -1903,6 +1911,21 @@ components:
...
+      type: object
+      description: Object modelling a document attribute request.
+      properties:
+        attributeId:
+          type: string
+          description: The attribute Id.          
+        value:
+          type: string
+          description: The attribute value.          
+    DocumentVersionAttributeRequestList:
```
Comment (melissapalmer) [inline doc/dokuti_api.yml]: Do we need a model for the list? as in DocumentVersionAttributeRequestList -- can a request/response not just be an array of DocumentAttributeValueRequest

File: src/main/java/za/co/grindrodbank/dokuti/document/DocumentControllerImpl.java [hunk 3]
```diff
@@ -218,19 +218,42 @@ private DocumentAttribute createOrUpdateDocumentAttribute(UUID documentVersionId
...
+        DocumentEntity documentEntity = documentService.findById(documentId);
+        UUID documentVersionId = documentEntity.getLatestDocumentVersion().getId();
+        if (documentInfoRequest.getName() != null) {
+            documentEntity.setName(documentInfoRequest.getName());
+        }
+        if (documentInfoRequest.getDescription() != null) {
+            documentEntity.setDescription(documentInfoRequest.getDescription());
+        }
+        if (documentInfoRequest.getTags() != null) {
+            addDocumentTags(documentId, documentInfoRequest.getTags());
```
Comment (melissapalmer) [inline src/main/java/za/co/grindrodbank/dokuti/document/DocumentControllerImpl.java]: Is there a need at all to worry about tags being removed?

File: src/main/java/za/co/grindrodbank/dokuti/document/DocumentControllerImpl.java [hunk 4]
```diff
@@ -218,19 +218,42 @@ private DocumentAttribute createOrUpdateDocumentAttribute(UUID documentVersionId
...
+        }
+
+        documentEntity = documentService.save(documentEntity);
+
+        if (documentInfoRequest.getAttributes() != null) {
+            documentInfoRequest.getAttributes().forEach(e -> {
+                DocumentAttributeRequest attrReq = new DocumentAttributeRequest();
+                attrReq.setValue(e.getValue());
+                createDocumentAttribute(documentId, documentVersionId, Integer.parseInt(e.getAttributeId()), attrReq);
```
Comment (melissapalmer) [inline src/main/java/za/co/grindrodbank/dokuti/document/DocumentControllerImpl.java]: Same type of question here: should removal of attributes be considered here at all?

File: doc/dokuti_api.yml [hunk 5]
```diff
@@ -1874,6 +1874,14 @@ components:
         description:
           type: string
           example: new description of the document
+        shares:
+          $ref:  '#/components/schemas/SharedObjects'  
+        unshares:
+          $ref:  '#/components/schemas/SharedObjects'
```
[the remark this answers is not in the document]
Comment (akorneyc) [reply doc/dokuti_api.yml]: shares - add permissions
unshares - remove permissions

File: doc/dokuti_api.yml [hunk 6]
```diff
@@ -1874,6 +1874,14 @@ components:
         description:
```
[the remark this answers is not in the document]
Comment (akorneyc) [reply doc/dokuti_api.yml]: this is part of https://redmine.bigbaobab.org/issues/2805

File: doc/dokuti_api.yml (hunk 1 above)
Comment (akorneyc) [reply doc/dokuti_api.yml]: attributeId is not included as it was used in url

File: doc/dokuti_api.yml (hunk 2 above)
Comment (akorneyc) [reply doc/dokuti_api.yml]: if we need setup several attributes we need list

File: src/main/java/za/co/grindrodbank/dokuti/document/DocumentControllerImpl.java (hunk 4 above)
Comment (akorneyc) [reply src/main/java/za/co/grindrodbank/dokuti/document/DocumentControllerImpl.java]: current implementations only add new attributes.  To remove attributes there are another endpoints

File: src/main/java/za/co/grindrodbank/dokuti/document/DocumentControllerImpl.java (hunk 3 above)
Comment (akorneyc) [reply src/main/java/za/co/grindrodbank/dokuti/document/DocumentControllerImpl.java]: to remove tags - there are different endpoints

File: doc/dokuti_api.yml (hunk 6 above)
[the remark this answers is not in the document]
Comment (melissapalmer) [reply doc/dokuti_api.yml]: No you have just added more fields to the swagger spec, there is a description field within spec that you can add to, to improve the API docs. Add to it explaining what is share, unshare differences for example.

File: doc/dokuti_api.yml (hunk 2 above)
Comment (melissapalmer) [reply doc/dokuti_api.yml]: The request can be an an array of items, you don't need a separate object to just hold an array do you?

Comment (akorneyc) [reply doc/dokuti_api.yml]: it will be like:
 "attributes": [
    {
      "attributeId": "string",
      "value": "string"
    }
  ]

**Pull request merged**